pandas/tools/tests/test_concat.py

```diff
@@ -1031,6 +1031,24 @@ def test_concat_invalid_first_argument(self):
         expected = read_csv(StringIO(data))
         assert_frame_equal(result, expected)
 
+    def test_concat_no_unnecessary_upcasts(self):
+        # fixes #13247
+
+        for pdt in [pd.Series, pd.DataFrame, pd.Panel, pd.Panel4D]:
+            dims = pdt().ndim
+            for dt in np.sctypes['float']:
+                dfs = [pdt(np.array([1], dtype=dt, ndmin=dims)),
+                       pdt(np.array([np.nan], dtype=dt, ndmin=dims)),
+                       pdt(np.array([5], dtype=dt, ndmin=dims))]
+                x = pd.concat(dfs)
+                self.assertTrue(x.values.dtype == dt)
+
+            for dt in (np.sctypes['int'] + np.sctypes['uint']):
+                dfs = [pdt(np.array([1], dtype=dt, ndmin=dims)),
+                       pdt(np.array([5], dtype=dt, ndmin=dims))]
```

Comment:
can you add some tests that have mixed dtypes in the input as well (e.g. mixed float32/64 and int types)